beluga/test/beluga/views/test_take_evenly.cpp

```diff
@@ -0,0 +1,71 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include "beluga/views/take_evenly.hpp"
+
+#include <range/v3/range/conversion.hpp>
+
+namespace {
+
+TEST(TakeEvenlyView, NoElements) {
+  const auto input = std::vector<int>{};
+  const auto output = input | beluga::views::take_evenly(0) | ranges::to<std::vector>;
+  ASSERT_EQ(output.size(), 0);
+}
+
+TEST(TakeEvenlyView, TakeZero) {
+  const auto input = std::vector<int>{1, 2, 3, 4};
+  const auto output = input | beluga::views::take_evenly(0) | ranges::to<std::vector>;
+  ASSERT_THAT(output, testing::ElementsAre(1, 2, 3, 4));
+}
+
+TEST(TakeEvenlyView, TakeOne) {
+  const auto input = std::vector<int>{1, 2, 3, 4};
+  const auto output = input | beluga::views::take_evenly(1) | ranges::to<std::vector>;
+  ASSERT_THAT(output, testing::ElementsAre(1));
+}
+
+TEST(TakeEvenlyView, TakeAll) {
+  const auto input = std::vector<int>{1, 2, 3, 4};
+  const auto output = input | beluga::views::take_evenly(10) | ranges::to<std::vector>;
+  ASSERT_THAT(output, testing::ElementsAre(1, 2, 3, 4));
+}
+
+TEST(TakeEvenlyView, TakeTwoFromFour) {
+  const auto input = std::vector<int>{1, 2, 3, 4};
+  const auto output = input | beluga::views::take_evenly(2) | ranges::to<std::vector>;
+  ASSERT_THAT(output, testing::ElementsAre(1, 4));
+}
```

Comment:
Depending on how one defines `max_size` this would be quite unexpected:

```
TEST(TakeEvenlyView, TakeThreeFromFour) {
  const auto input = std::vector<int>{1, 2, 3, 4};
  const auto output = input | beluga::views::take_evenly(3) | ranges::to<std::vector>;
  ASSERT_THAT(output, testing::ElementsAre(1, 2, 3, 4));
}
```

Reply:
Added documentation in https://github.com/Ekumen-OS/beluga/commit/5049410c78a383d7a1896c3ee9754a80614ee26f mentioning that the last element is always included.

This is the nav2_amcl implementation: [nav2_amcl/src/sensors/laser/beam_model.cpp#L72](https://github.com/ros-planning/navigation2/blob/e2c8555ad352d69aa3da1e670b888d00f6eaeb3a/nav2_amcl/src/sensors/laser/beam_model.cpp#L72)

Reply:
> Added documentation in [5049410](https://github.com/Ekumen-OS/beluga/commit/5049410c78a383d7a1896c3ee9754a80614ee26f) mentioning that the last element is always included.

Is it, though? This is one of the test cases, and the last element is not included:

```
TEST(TakeEvenlyView, TakeThreeFromSix) {
  const auto input = std::vector<int>{1, 2, 3, 4, 5, 6};
  const auto output = input | beluga::views::take_evenly(3) | ranges::to<std::vector>;
  ASSERT_THAT(output, testing::ElementsAre(1, 3, 5));
}
```
The first element is always included; the last will only be if $(s - 1) / (c - 1)$ divides exactly. My example above looked weird because  the resulting `step` after division was $1$. Making it a bit more extreme, this example still returns 10 elements instead of 6 for that very same reason:

```
TEST(TakeEvenlyView, CommentTakeFiveFromNine) {
  const auto input = std::vector<int>{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  const auto output = input | beluga::views::take_evenly(6) | ranges::to<std::vector>;
  ASSERT_THAT(output, testing::ElementsAre(1, 2, 3, 4, 5, 6, 7, 8, 9, 10));
}
```

Look at 27d863d for an alternative implementation of `take_evenly` that I experimented with that uses floating point to try to keep things even-ish even when the division is not exact.

Reply:
Neat! Cherry-picked here with small changes: https://github.com/Ekumen-OS/beluga/pull/271/commits/ed57a9830be8ecb46e9dea5145ee4c314557fc47

The floating point conversions don't seem necessary based on the test results. Let me know if I'm missing something.